### DISCUSSION TITLE
Cake.exe : error CS0234: The type or namespace name 'DotNet' does not exist in the namespace 'Cake.Common.Tools'

### DISCUSSION THREAD

**Discussion by dhanrajshimpi123** [cake-build/cake#3850] [Extension Q&A]
Hi All

We are getting error Cake.exe : error CS0234: The type or namespace name 'DotNet' does not exist in the namespace 'Cake.Common.Tools'

We are trying to build dotnetcore3.1 app using CAKE 2.1.0. Also we have install .NET 6 & .NET CORE 3.1 on machine where we run Build.Cake script. 

Here is the code snippet of build.cake script.

#addin nuget:?package=Cake.Services&version=1.0.1
#addin nuget:?package=Cake.Powershell&version=1.0.1
#tool nuget:?package=OctopusTools&version=6.13.1
#tool dotnet:?package=Cake.Tool&version=2.1.0

#region Namespaces

using Newtonsoft.Json;
using Cake.Common.Tools.OctopusDeploy;
using Cake.Common.Tools.DotNet.Build;
using Cake.Common.IO;
using Cake.Core.IO;
using Cake.Common.Tools.NuGet.Restore;

Task("Build")
	.IsDependentOn("Restore-NuGet-Packages")
	.Does<BuildContext>((context) =>
	{
		var buildSettings = new DotNetBuildSettings()
		{
			Configuration = context.Configuration,
			NoRestore = true
		};

		if (!string.IsNullOrEmpty(project.Runtime))
		{
			buildSettings.Runtime = project.Runtime;
		}

		DotNetBuild(project.FullPath, buildSettings);
		
		var projectDir = new DirectoryPath(context.DistDirectory);

		var publishSettings = new DotNetPublishSettings
		{
			Configuration = context.Configuration,
			OutputDirectory = projectDir,
			ArgumentCustomization = args => args.Append("/p:PublishProfile=MyPublishProfile"),
		};				

		DotNetPublish(project.FullPath, publishSettings);
		Information("Build done.");
		
	});

==================
Error Log: 

PS D:\WKH-LRP\CloudServices\Builds\CIBS\EngagementEngine> .\build.ps1
Preparing to run build script...
Running build script...
Tool Cake.Tool is already installed, with required version.
Cake.exe : error CS0234: The type or namespace name 'DotNet' does not exist in the namespace 'Cake.Common.Tools' (are you missing an assembly reference?)
At D:\WKH-LRP\CloudServices\Builds\CIBS\EngagementEngine\build.ps1:241 char:1
+ &$CAKE_EXE $cakeArguments
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (error CS0234: T...bly reference?):String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(131,27): error CS0246: The type or namespace name 'DotNetBuildSettings' could not be found (are you missing a using directive or an assembly reference?)
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(142,3): error CS0103: The name 'DotNetBuild' does not exist in the current context
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(146,29): error CS0246: The type or namespace name 'DotNetPublishSettings' could not be found (are you missing a using directive or an assembly reference?)
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(153,3): error CS0103: The name 'DotNetPublish' does not exist in the current context
Error: Error(s) occurred when compiling build script:
error CS0234: The type or namespace name 'DotNet' does not exist in the namespace 'Cake.Common.Tools' (are you missing an assembly reference?)
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(131,27): error CS0246: The type or namespace name 'DotNetBuildSettings' could not be found (are you missing a using directive or an assembly reference?)
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(142,3): error CS0103: The name 'DotNetBuild' does not exist in the current context
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(146,29): error CS0246: The type or namespace name 'DotNetPublishSettings' could not be found (are you missing a using directive or an assembly reference?)
D:/WKH-LRP/CloudServices/Builds/CIBS/EngagementEngine/build.cake(153,3): error CS0103: The name 'DotNetPublish' does not exist in the current context

**Comment by nils-a** [comment]
@dhanrajshimpi123 Sorry to hear to you're having problems.

Can you please run `.\build.ps1 --version` to verify that you are using Cake 2.1.0?

**Reply by devlead** [reply]
I would suspect you're using an earlier version of Cake.

A few notes
```cake
#tool dotnet:?package=Cake.Tool&version=2.1.0
```
Will install `Cake.Tool` in the tools folder, as you're already running Cake this is a bit redundant, it will not use Cake.Tool 2.1 as Cake already executing.

You can read more about bootstrapping Cake here
https://cakebuild.net/docs/getting-started/setting-up-a-new-scripting-project

**Reply by dhanrajshimpi123** [reply]
> @dhanrajshimpi123 Sorry to hear to you're having problems.
> 
> Can you please run `.\build.ps1 --version` to verify that you are using Cake 2.1.0?

@nils-a  I have run the '.\build.ps1 --version' 

It showing 1.3.0 

Also we are able to see this version in package.config  

package id="Cake" version="1.3.0" 

If we changed this package version to 2.1 then we will get error below error.

PS D:\WKH-LRP\CloudServices\Builds\CIBS\EngagementEngine> .\build.ps1
Preparing to run build script...
nuget.exe : Unable to find version '2.1.0' of package 'Cake'.
At line:1 char:1
+ &"D:\WKH-LRP\CloudServices\Builds\CIBS\EngagementEngine\tools\nuget.e ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Unable to find ...package 'Cake'.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
  C:\Users\administrator\.nuget\packages\: Package 'Cake.2.1.0' is not found on source 'C:\Users\administrator\.nuget\packages\'.
  https://pkgs.dev.azure.com/cake-build/Cake/_packaging/cake/nuget/v3/index.json: Package 'Cake.2.1.0' is not found on source 'https://pkgs.dev.azure.com/cake-build/Cake/_packaging/cake/nuget/v3/index.json'.
An error occurred while restoring NuGet tools.
At D:\WKH-LRP\CloudServices\Builds\CIBS\EngagementEngine\build.ps1:175 char:9
+         Throw "An error occurred while restoring NuGet tools."
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (An error occurr...ng NuGet tools.:String) [], RuntimeException
    + FullyQualifiedErrorId : An error occurred while restoring NuGet tools.

**Reply by nils-a** [reply]
Yes, that confirms my initial assumption: You are using Cake 1.3.0 and not Cake 2.1.0 - hence the original error of `error CS0234: The type or namespace name 'DotNet' does not exist in the namespace 'Cake.Common.Tools'`.

As to the "new" error: `nuget.exe : Unable to find version '2.1.0' of package 'Cake'.` - this is to be expected. Cake 1.3.0 was the last version to be shipped in the .NET Framework Runner. Newer versions can only be used, using the Cake .NET Tool (or Cake Frosting).

If this is a new Cake project, I suggest you have a look at https://cakebuild.net/docs/getting-started/setting-up-a-new-scripting-project, if this is an older project that was set up before Cake 2.0.0, I suggest you have a look at https://cakebuild.net/docs/getting-started/upgrade#cake-1.x-to-cake-2.0.

Also, as @devlead already mentioned: Having `#tool dotnet:?package=Cake.Tool&version=2.1.0` in your script is probably not what you wanted, since it won't be used.
Additionally, all the `using Cake.*` should not be needed.
